roles/pulp_database_config/tasks/main.yml

```diff
@@ -1,7 +1,184 @@
+          Run pulp_database_config(or pulp_services, or pulp_all_services) against a host that does
+          have the key without setting `pulp_database_config_host`, or set
+          `pulp_database_config_host` to a host that has the key.
+      when:
+        - hostvars['localhost']['pulp_database_config_host'] is defined
+
+    - name: Pick & set the sole host used to run pulp_database_config, if any hosts have the key
+      set_fact:
+        pulp_database_config_host: "{{ item.item }}"
+      run_once: True
+      delegate_to: localhost
+      delegate_facts: yes
+      with_items: "{{ __pulp_db_fields_key_stat.results }}"
+      # We break the loop if the variable is already set, from this loop or by the user.
+      # We only set the variable if its a host where they exists.
+      when:
+        - hostvars['localhost']['pulp_database_config_host'] is not defined
+        - item.item in hostvars['localhost']['__pulp_database_config_hosts']
+        - item.stat.exists
+
+    - name: Fail if some hosts in the play have the key, but no hosts running pulp_database_config do
+      assert:
+        that: >
+          (__pulp_db_fields_key_stat.results | selectattr('stat.exists', 'equalto', True) | list | count == 0) or
+          (__pulp_db_fields_key_stat.results | selectattr('stat.exists', 'equalto', True) | list | selectattr('item', 'in', hostvars['localhost']['__pulp_database_config_hosts']) | list | count >= 1)
+        fail_msg: >
+          pulp_installer cannot continue because none of the hosts you are running
+          pulp_database_config (or pulp_services, or pulp_all_services) against have the
+          database fields encryption key ({{ __pulp_db_fields_key_path }}) but some other hosts
+          in the ansible playbook run
+          ({{ groups['all'] | difference(hostvars['localhost']['__pulp_database_config_hosts']) }}) do.
+          Run pulp_database_config against a host that does already have the key, or set
+          `pulp_database_config_host` to a host that has the key (which is an
+          alternative way to specify which host pulp_database_config runs against.)
+      when:
+        - hostvars['localhost']['pulp_database_config_host'] is not defined
+
+    - name: Verify that all the hosts' keys are identical to the sole host's key
+      assert:
+        that: >
+          item['stat']['checksum']
+          == __pulp_db_fields_key_stat.results | selectattr('item', 'equalto', hostvars['localhost']['pulp_database_config_host']) | map(attribute='stat.checksum') | first
+        fail_msg: >
+          Pulp_installer cannot proceed because the database fields key
+          ({{ __pulp_db_fields_key_path }}) differs across hosts.
+          You can manually resolve this by downloading the correct key
+          from a host to the ansible management node, and then specifying the variable
+          `pulp_db_fields_key` to its filepath.
```

Comment:
@fao89 @mdellweg @mikedep333 , is it always a solution, or is there a way when the user can be totally doomed and used different keys from different nodes to create different db records?  I think we had this discussion in some chat, not sure what was the outcome.
Do we want to ask the reporter of the issue or QE folks to test this scenario? Creating remotes from different nodes in HA setup (is it possible) before the fix, and then run again with the fix.

Reply:
Honestly, I don't know.

I suspect that some fields in the database were encrypted with one key, and others encrypted with a different key.

I suspect that the only way to properly fix it would be to add key rotation logic, that looks for encrypted fields with a list of old keys, and converts them to being encrypted with the new key. This would be application logic that the installer would adapt to.

That said, the best approach is to ask them to test it like you mentioned.